### ISSUE TITLE
Interactivity does not work in Jupyter Lab

### ISSUE THREAD

**Issue by lschr** [matplotlib/jupyter-matplotlib#66]
Everything works fine in Jupyter Notebook, but not in Jupyter Lab (version 0.34.7).

![screenshot](https://user-images.githubusercontent.com/13418822/45416986-6a953580-b681-11e8-8d4e-d36942fc0836.png)

- Styling of the buttons does not seem to work
- The buttons have no functionality; e.g. pressing the panning button does not enable panning

On the Javascript console I can the following error whenever the mouse pointer enters the area of the figure, which may or may not be related:

![screenshot_20180912_130449](https://user-images.githubusercontent.com/13418822/45421134-6f131b80-b68c-11e8-89e0-5048c568f859.png)

**Comment by feanor12** [comment]
I had a similar problem when I was using a notebook which had output cells with plots generated by (%matplotlib notebook)

I fixed it by:
1. delete all output cells
2. save notebook
3. reload page
4. restart kernel 
5. run with %matplotlib widget

I think that %matplotlib notebook injects some kind of javascript that kills something inside jupyter lab.

**Comment by lschr** [comment]
@feanor12 Thanks, it looks like you are right. After restarting and clearing all output, the buttons work (except for the "power" button, which has already been reported in #61) and also the mouse pointer coordinates are shown properly in the bottom right corner.

The buttons still look Windows 95-ish, though (not sure, if this is a bug).

**Comment by feanor12** [comment]
So maybe this is the cause for #62 as well.

**Comment by MerlinPendragon** [comment]
@feanor12 Yes!

**Comment by pdeitel** [comment]
Is this supposed to work with Matplotlib FuncAnimations too? All I get is a blank output and the FuncAnimation itself--which works just fine in IPython in a Terminal window--does not call the update function I defined for each frame.  Thanks in advance for any suggestions.

**Comment by tacaswell** [comment]
@pdeitel that is extremely odd.  I am a bit worried that your generator is raising someplace and the exceptions are getting snarfed incorrectly.  Internally, the animation module homgonizes all the input to generators (an integer gets turned into `range`).

**Comment by tacaswell** [comment]
> I think that %matplotlib notebook injects some kind of javascript that kills something inside jupyter lab.

Yes, all of the js to make the plot interactive is inserted by the notebook backend (which is why it does not work in jlab as they have closed that convenient but large security hole).

**Comment by goiosunsw** [comment]
Additionally to @feanor12's recipe I had to close all tabs on the right hand side (open notebooks). Probably their javascript was still interfering with ipympl

**Comment by amueller** [comment]
Seems related so adding here:
If I try
```
% matplotlib widget
```
in a notebook that had ``%matplotlib notebook`` before, I get
```
Warning: Cannot change to a different GUI toolkit: widget. Using notebook instead.
```

I understand this to mean "use ``% matplotlib notebook`` instead" which seems bad advice. Maybe that's not what it means?
I know I'm doing it wrong, but I think the warning is misleading.

**Comment by dopplershift** [comment]
It means once you’ve run `%matplotlib notebook`, you can’t later switch to `%matplotlib widget` (because the two will use different event loops). You need to only ever use one or the other while the kernel is running.

Open to suggestions to language that would make that more clear, but unfortunately that’s exactly what the existing message says to me (but then again I’ve been doing this for way too long).

**Comment by pdeitel** [comment]
How about adding something like

“Cannot use %matplotlib notebook and %matplotlib widget in the same notebook because they use different event loops”?

Paul J. Deitel, CEO
Deitel & Associates, Inc.
Oracle Java Champion
Microsoft C# MVP 2012-2014
Facebook.com/DeitelFan
@deitel

Sent from my iPhone

On Jan 29, 2019, at 6:07 AM, Thomas Aarholt <notifications@github.com> wrote:

The warning should probably explain why it cannot change backend, or at least how the user can use a different one.

My suggestion below is verbose, but I think 50% of basic users have probably encountered this error message and then become very confused or worried.

How about changing

Warning: Cannot change to a different GUI toolkit: widget. Using notebook instead.
to

Warning: Cannot change GUI toolkit from notebook to widget. 
Only the inline and *one* interactive toolkit can be used together. 
Restart kernel to use a different toolkit.

**Comment by tobiasjj** [comment]
Probably another solution that works for me in both, notebook and jupyterlab (up to version 0.35.4):

```
import matplotlib
matplotlib.use('module://ipympl.backend_nbagg')

from matplotlib import pyplot as plt
fig, ax = plt.subplots()
ax.plot([1,2,3], [4,3,2])
fig.show()
```

**Comment by yt87** [comment]
I still see the behaviour as described by @MerlinPendragon in his original post issue #62  on Firefox. On Chromium the example runs fine. 

I am on Fedora 29.
```
$ conda list | grep jupyter                       
jupyter                   1.0.0                      py_1    conda-forge
jupyter_client            5.2.4                      py_0    conda-forge
jupyter_console           6.0.0                      py_0    conda-forge
jupyter_core              4.4.0                      py_0    conda-forge
jupyterlab                0.35.3                   py37_0  
jupyterlab_launcher       0.13.1                     py_2    conda-forge
jupyterlab_server         0.2.0                    py37_0

$ rpm -q firefox
firefox-64.0-4.fc29.x86_64

$ jupyter labextension list                       
JupyterLab v0.35.3
Known labextensions:
   app dir: /mnt/sdc1/home/gtrojan/miniconda3/envs/devel/share/jupyter/lab
        @deathbeds/jupyterlab-fonts v0.6.0  enabled  OK
        @jupyter-widgets/jupyterlab-manager v0.38.1  enabled  OK
        @pyviz/jupyterlab_pyviz v0.6.4  enabled  OK
        jupyter-matplotlib v0.3.0  enabled  OK
        jupyterlab_vim v0.10.1  enabled  OK
```

**Comment by steviestevepy** [comment]
> import matplotlib
> matplotlib.use('module://ipympl.backend_nbagg')

This solution worked form me, only after restarting the kernel and jupyterlab.
Looks like the widget function has bugs. After manipulating the graph and then hitting the button to turn off interactivity the graph goes all fuzzy and turns to black and white.
I don't know what the issue is for that

**Comment by flutefreak7** [comment]
What about:

Matplotlib is currently configured to use the 'notebook' backend which is suitable for the classic Jupyter Notebook and derivatives (and does not work in JupyterLab).  Matplotlib does not support changing from the 'notebook' backend to the 'widget' backend while the kernel is running. To use the 'widget' backend, which is supported by JupyterLab, it is recommended that you restart your kernel and clear all output cells. See (insert helpful link to docs) for details.

**Comment by drunkwretch** [comment]
> It means once you’ve run `%matplotlib notebook`, you can’t later switch to `%matplotlib widget` (because the two will use different event loops). You need to only ever use one or the other while the kernel is running.
> 
> Open to suggestions to language that would make that more clear, but unfortunately that’s exactly what the existing message says to me (but then again I’ve been doing this for way too long).

right, just restart the kernal and run `%matplotlib widget`

**Comment by Nestak2** [comment]
I have a similar issue - I have installed everything as [required here](https://github.com/matplotlib/jupyter-matplotlib). When I do an interactive plot in Jupyter Lab it is run without an error message, but the plot is just not displayed. Instead I have the output from the screenshot. What should I do?
![Screenshot 2019-08-03 at 14 26 48](https://user-images.githubusercontent.com/16193553/62412469-c8fd3680-b5fa-11e9-91ce-495b1e18abf5.png)

**Comment by cottrell** [comment]
None of the suggestions work currently. This was working recently.

**Comment by martinRenou** [comment]
@cottrell can you tell us how you installed ipympl? You might have missed an installation step.

I think this issue should be closed. ipympl does work in JupyterLab now.

**Comment by jasongrout** [comment]
> I think this issue should be closed. ipympl does work in JupyterLab now.

As you wish. @cottrell, feel free to open a new issue with more comprehensive information about your environment and a reproducible test case.

**Comment by cottrell** [comment]
Interesting ... I just tried at home on a Linux system (again following https://stackoverflow.com/questions/50149562/jupyterlab-interactive-plot) and everything works again. 

Will try again on the work setup ... likely some interaction or version issue.

**Comment by Chang-Hsinyi** [comment]
Hi, 

I use %matplotlib ipympl, all the functions can work, exclude resize function at the below right corner. 
Does anyone know what the problem is, and how can I fix it? thanks

**Comment by ijpulidos** [comment]
I'm still facing the issue here. All I get is that `Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …`

Jupyterlab 2.1.4
ipympl 0.5.6

You can find the `Pipfile` and `Pipfile.lock` files in https://gist.github.com/ijpulidos/81edbe253bdbbe270e94a0eff75f34c0 (using pipenv to manage the environment)

**Comment by ijpulidos** [comment]
Ok, had to manually run `jupyter labextension install @jupyter-widgets/jupyterlab-manager@2` for them to work. Not really sure why or how all of this works, but that did it for me.

**Comment by anoetzli** [comment]
> Ok, had to manually run `jupyter labextension install @jupyter-widgets/jupyterlab-manager@2` for them to work. Not really sure why or how all of this works, but that did it for me.

Thank you - this solution worked for me on 
Jupyerlab 2.2.4
ipympl 0.5.7
